File: 278. First Bad Version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        use two pointers
        """
        head, tail = 1, n
        while head < tail:
            mid = (head + tail) / 2
            if isBadVersion(mid):
                tail = mid
            else:
                head = mid + 1
        return head

File: 278. First Bad Version/278. First Bad Version.js
/**
 * Definition for isBadVersion()
 * 
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
        var head = 1, tail = n;
        while (head < tail){
            var mid = parseInt((tail + head) / 2);
            /**
             * same with python use parseInt cut the int part
             */
            if (isBadVersion(mid))
                tail = mid;
            else
                head = mid + 1;
        }
        return head;
    };
};
